hal/modem_api.h

```diff
@@ -0,0 +1,88 @@
+/* mbed Microcontroller Library
```

Comment:
> I dont understand this commit. The modem code looks to be very similar and only implemented for the C027 ublox modem. What if I bought https://www.u-blox.com/en/2g-and-3g-modules and hooked it upto a K64F? How do I make use of these drivers?

This is the fundamental disconnect in mbed OS - on-board devices work totally differently to off-board devices.

This driver is intended for on-board devices connected via an on-board UART, and uses a modem HAL to access their meta-control functions. As an on-board device driver it is easy for applications to use - they can use this class without knowing anything about the platform. A platform just needs to implement DEVICE_MODEM to provide those functions.

Off-board devices would need a different driver, or a variation of this one. And applications would need to know where to find that driver and how to configure it.

It would be nice if there was general unification between on-board and off-board devices in mbed OS, but in the absence of that, we decided to go for an on-board driver, as that's what the two initial targets have.